calico_containers/adapter/datastore.py

```diff
@@ -444,36 +463,112 @@ def get_bgp_peers(self, version):
+        for node in nodes:
+            if node.value is None:
+                continue
+            try:
+                data = json.loads(node.value)
+                data["ip"] = IPAddress(data["ip"])
+                peers.append(data)
+            except BaseException:
+                pass
+        return peers
+
+    @handle_errors
+    def add_bgp_peer(self, version, ip, as_num):
+        """
+        Add a global BGP Peer.
+
+        If the peer already exists then do nothing.
+
+        :param version: "v4" for IPv4, "v6" for IPv6
+        :param ip: The IP address to add. (an IPAddress)
+        :return: Nothing
+        """
+        assert version in ("v4", "v6")
+        assert isinstance(ip, IPAddress)
+        bgp_peer_path = BGP_PEER_PATH % {"version": version,
+                                         "peer_ip": str(ip)}
+        data = {"as_num": as_num, "ip": str(ip)}
+        self.etcd_client.write(bgp_peer_path, json.dumps(data))
+
+    @handle_errors
+    def remove_bgp_peer(self, version, ip):
+        """
+        Delete a global BGP Peer.
+
+        :param version: "v4" for IPv4, "v6" for IPv6
+        :param ip: The IP address to delete. (an IPAddress)
+        :return: Nothing
+        """
+        assert version in ("v4", "v6")
+        assert isinstance(ip, IPAddress)
+        bgp_peer_path = BGP_PEER_PATH % {"version": version,
+                                         "peer_ip": str(ip)}
+        try:
+            self.etcd_client.delete(bgp_peer_path)
+        except EtcdKeyNotFound:
+            # Re-raise with a better error message.
+            raise KeyError("%s is not a configured peer." % ip)
+
+    @handle_errors
+    def get_node_bgp_peers(self, hostname, version):
         """
-        Retrieve all the keys in a path and create a reverse dict
-        values -> keys
+        Get the configured BGP Peers
 
-        :param path: The path to get the keys from.
-        :return: dict of {<values>: <etcd key>}
+        :param version: "v4" for IPv4, "v6" for IPv6
+        :return: List of dictionaries:
+                    [
+                      {
+                        "ip": <peer ip address>,
+                        "as_num": <peer as num>
+                      },
+                      ...
+                    ]
         """
+        assert version in ("v4", "v6")
+        bgp_peers_path = HOST_BGP_PEERS_PATH % {"hostname": hostname,
+                                                "version": version}
 
         try:
-            nodes = self.etcd_client.read(path).children
+            nodes = self.etcd_client.read(bgp_peers_path).children
         except EtcdKeyNotFound:
             # Path doesn't exist.
-            return {}
-        else:
-            values = {}
-            for child in nodes:
-                value = child.value
-                if value:
-                    values[value] = child.key
-            return values
+            return []
+
+        peers = []
```

Comment:
isn't this just copy-and-paste from `get_bgp_peers` with a different path?  Can you refactor to call a helper function with all the common code?